src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,244 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
+  rcases hz with ⟨c, eq1⟩,
+  rw [finsupp.total_apply, finsupp.sum] at eq1,
+  obtain ⟨⟨_, N, rfl⟩, hN⟩ := is_localization.exist_integer_multiples_of_finset (submonoid.powers f)
+    (c.support.image c),
+  choose acd hacd using hN,
+  have prop1 : ∀ i, i ∈ c.support → c i ∈ finset.image c c.support,
+  { intros i hi, rw finset.mem_image, refine ⟨_, hi, rfl⟩, },
+
+  refine ⟨c, N, acd, _⟩,
+  rw [← eq1, smul_sum, map_sum, ← sum_attach],
+  congr' 1,
+  ext i,
+  rw [_root_.map_mul, hacd, (classical.some_spec i.1.2).2, smul_eq_mul, smul_mul_assoc],
+  refl
+end
+
+lemma no_intersection :
+  (x.1.as_homogeneous_ideal.to_ideal ∩ submonoid.powers f : set A) = ∅ :=
+begin
+  by_contra rid,
+  rw [←ne.def, set.ne_empty_iff_nonempty] at rid,
+  choose g hg using rid,
+  obtain ⟨hg1, ⟨k, rfl⟩⟩ := hg,
+  by_cases k_ineq : 0 < k,
+  { erw x.1.is_prime.pow_mem_iff_mem _ k_ineq at hg1,
+    exact x.2 hg1 },
+  { erw [show k = 0, by linarith, pow_zero, ←ideal.eq_top_iff_one] at hg1,
+    apply x.1.is_prime.1,
+    exact hg1 },
+end
+
+lemma carrier_ne_top :
+  carrier f_deg x ≠ ⊤ :=
+begin
+  have eq_top := no_intersection x,
+  classical,
+  contrapose! eq_top,
+  obtain ⟨c, N, acd, eq1⟩ := mem_carrier.clear_denominator _ x ((ideal.eq_top_iff_one _).mp eq_top),
+  rw [algebra.smul_def, subtype.val_eq_coe, subring.coe_one, mul_one] at eq1,
+  change localization.mk (f ^ N) 1 = mk (∑ _, _) 1 at eq1,
+  simp only [mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  erw [mul_one, mul_one] at eq1,
+  simp only [← subtype.val_eq_coe] at eq1,
+  refine set.ne_empty_iff_nonempty.mpr ⟨f^N * f^M, eq1.symm ▸ mul_mem_right _ _
+    (sum_mem _ (λ i hi, mul_mem_left _ _ _)), ⟨N+M, by rw pow_add⟩⟩,
+  generalize_proofs h,
+  exact (classical.some_spec h).1,
+end
+
+/--The function between the basic open set `D(f)` in `Proj` to the corresponding basic open set in
+`Spec A⁰_f`. This is bundled into a continuous map in `Top_component.forward`.
+-/
+def to_fun : (Proj.T| (pbo f)) → (Spec.T (A⁰_ f_deg)) := λ x,
+⟨carrier f_deg x, carrier_ne_top f_deg x, λ x1 x2 hx12, begin
+  haveI : decidable_eq (away f) := classical.dec_eq _,
+  rcases ⟨x1, x2⟩ with ⟨⟨x1, hx1⟩, ⟨x2, hx2⟩⟩,
+  induction x1 using localization.induction_on with data_x1,
+  induction x2 using localization.induction_on with data_x2,
+  rcases ⟨data_x1, data_x2⟩ with ⟨⟨a1, _, ⟨n1, rfl⟩⟩, ⟨a2, _, ⟨n2, rfl⟩⟩⟩,
+  rcases mem_carrier.clear_denominator f_deg x hx12 with ⟨c, N, acd, eq1⟩,
+  simp only [degree_zero_part.mul_val, localization.mk_mul, mul_one, algebra.smul_def] at eq1,
+  change localization.mk (f ^ N) 1 * mk _ (⟨f ^ n1 * f ^ n2, _⟩) = mk (∑ _, _) _ at eq1,
+  rw [mk_mul, one_mul] at eq1,
+  simp only [mk_mul, mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  rw [submonoid.coe_one, mul_one] at eq1,
+  simp only [← subtype.val_eq_coe, submonoid.coe_mul] at eq1,
```

Comment:
If your proof uses `val_eq_coe`, that's usually an indication that one of your lemma statements used `.1` where it should have used `coe`

Reply:
done